orttraining/orttraining/training_ops/cpu/op_gradients.cc

```diff
@@ -74,80 +84,108 @@ Status SoftmaxGrad<T>::Compute(OpKernelContext* context) const {
+    D = TensorShape(transposed_input_dims).SizeFromDimension(rank - 1);
 
-  return Status::OK();
-}
-
-ONNX_OPERATOR_KERNEL_EX(
-    LogSoftmaxGrad,
-    kMSDomain,
-    1,
-    kCpuExecutionProvider,
-    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
-    LogSoftmaxGrad<float>);
-
-template <typename T>
-Status LogSoftmaxGrad<T>::Compute(OpKernelContext* context) const {
-  auto& dY = *context->Input<Tensor>(0);
-  auto& Y = *context->Input<Tensor>(1);
-  const TensorShape input_shape{Y.Shape()};
-  auto& dX = *context->Output(0, Y.Shape());
+    // Allocate a temporary tensor to hold transposed input
+    auto temp_input0 = Tensor::Create(Y.DataType(), TensorShape(transposed_input_dims), alloc);
 
-  auto axis = HandleNegativeAxis(axis_, Y.Shape().NumDimensions());
+    // Perform the transpose
+    ORT_RETURN_IF_ERROR(Transpose::DoTranspose(permutation, Y, *temp_input0));
+    transposed_Y = std::move(temp_input0);
 
-  size_t N = input_shape.SizeToDimension(axis);
-  size_t D = input_shape.SizeFromDimension(axis);
+    auto temp_input1 = Tensor::Create(Y.DataType(), TensorShape(transposed_input_dims), alloc);
+    ORT_RETURN_IF_ERROR(Transpose::DoTranspose(permutation, dY, *temp_input1));
+    transposed_dY = std::move(temp_input1);
 
-  if (N == 0) {
-    return Status::OK();
+    // Allocate memory for the intermediate output
+    intermediate_output = Tensor::Create(dX.DataType(), TensorShape(transposed_input_dims), alloc);
   }
 
+  const int n = gsl::narrow_cast<int>(N);
   const int d = gsl::narrow_cast<int>(D);
   const int nd = gsl::narrow_cast<int>(N * D);
+  const float* Ydata = is_transpose_required ? transposed_Y->template Data<T>() : Y.template Data<float>();
+  const float* dYdata = is_transpose_required ? transposed_dY->template Data<T>() : dY.template Data<float>();
+  float* dXdata = is_transpose_required ? intermediate_output->template MutableData<T>() : dX.template MutableData<float>();
 
-  const float* Ydata = Y.template Data<float>();
-  const float* dYdata = dY.template Data<float>();
-  float* dXdata = dX.template MutableData<float>();
-
-  std::vector<float> eY(nd);
-  float* eYdata = eY.data();
-
-  // dX_ai = d(log Y_ai) - [sum_j d(log Y_aj)] exp(log Y_ai)
   gsl::copy(gsl::make_span(dYdata, nd), gsl::make_span(dXdata, nd));
-  math::Exp<float, CPUMathUtil>(nd, Ydata, eYdata, nullptr);
-  for (size_t i = 0; i < N; ++i) {
-    float sdY;
-    math::Sum<float, CPUMathUtil>(d, dYdata + i * d, &sdY, nullptr, nullptr);
-    math::Axpy<float, CPUMathUtil>(d, -sdY, eYdata + i * d, dXdata + i * d, nullptr);
+  if (is_logsoftmaxgrad_) {
+    std::vector<float> eY(nd);
+    float* eYdata = eY.data();
+
+    // dX_ai = d(log Y_ai) - [sum_j d(log Y_aj)] exp(log Y_ai)
+    gsl::copy(gsl::make_span(dYdata, nd), gsl::make_span(dXdata, nd));
+    math::Exp<float, CPUMathUtil>(nd, Ydata, eYdata, nullptr);
+    for (size_t i = 0; i < N; ++i) {
```

Comment:
Why not parallelizing this loop too?

Reply:
This is CPU gradient operation. This code exists only as a fallback.